### ISSUE TITLE
The difference in batch size leads to different results in LongBench testing

### ISSUE THREAD

**Issue by Felixvillas** [jy-yuan/KIVI#18]
When I took the LongBench test with `batch_size=1`, I got the same results as in Table 4 of the paper. However, when I increased the batch size, the results were worse than with `batch_size=1`. For example, the result with `model=Llama2-7B, dataset=TREC, batch_size=1, group_size=32, residual_length=128, k_bits=2, v_bits=2` was 66.0, but it was 57.0 with `batch_size=2`.

Moreover, the results with 16 bits remain consistent across different `batch_size`, which indicates that the Llama model's transformer supports it.

Is there a shared state or cache within a batch that could be causing this issue?

Hope to hear from you soon.

**Comment by zirui-ray-liu** [comment]
Thank you for the information.

> Is there a shared state or cache within a batch that could be causing this issue?

No, there is no sharing across batch size. In KIVI, for key cache, scale and offset factor are shared between different tokens within the same batch; for value cache, scale and offset factor are shared between different channels within the same embedding.

It is a bug. My guess is because of padding, which pad different request into the same length. I will try to fix it

**Comment by Felixvillas** [comment]
> Thank you for the information.
> 
> > Is there a shared state or cache within a batch that could be causing this issue?
> 
> No, there is no sharing across batch size. In KIVI, for key cache, scale and offset factor are shared between different tokens within the same sequence; for value cache, scale and offset factor are shared between different channels within the same token embedding.
> 
> It is a bug. My guess is because of padding, which pad different request into the same length. The current [pred_long_bench.py](https://github.com/jy-yuan/KIVI/blob/main/pred_long_bench.py) does not support inference with batch size > 1. Can you share your test file & script with me? So I can have a try.

Thank you for your response! I also think that the accuracy issue is due to padding. Therefore, I believe padding all input prompts to 4096, which corresponds to the Llama model, even when the batch size is 1, should solve the problem. Below is my original code:
`# LLaMA model with KIVI
import torch
import os
from tqdm import tqdm
# os.environ["CUDA_VISIBLE_DEVICES"]="4"
from transformers import LlamaConfig, AutoTokenizer
import time
import get_model_name
from openpyxl import Workbook
import json
import argparse
from datasets import load_dataset
import json

parser = argparse.ArgumentParser(description='Process some parameters.')
parser.add_argument('--bits', nargs='+', type=str, default=['2', '16'],
                    help='List of bits')
parser.add_argument('--residual_length', nargs='+', type=str, default=['32', '128'],
                    help='List of residual length')
parser.add_argument('--group_size', nargs='+', type=str, default=['32'],
                    help='List of group size')
parser.add_argument('--batch_size', nargs='+', type=str, default=['2', '4', '8', '16', '32', '64', '96', '128', '256', '384'],
                    help='List of batch size')
parser.add_argument('--datasets', nargs='+', type=str, default=["triviaqa", "qasper", "trec", "samsum", "lcc", "repobench-p", "qmsum", "multi_news"],
                    help='List of datasets')
parser.add_argument('--model_name_or_path', type=str, default='Llama-2-7b-hf',
                    help='Model name or path')
parser.add_argument('--cache_dir', type=str, default='./cached_models',
                    help='Cache directory')
args = parser.parse_args()

def post_process(response, model_name):
    if "xgen" in model_name:
        response = response.strip().replace("Assistant:", "")
    elif "internlm" in model_name:
        response = response.split("<eoa>")[0]
    return response

def pred2jsonl(out_path, dataset, preds):
    os.makedirs(os.path.dirname(out_path), exist_ok=True)
    if not os.path.exists(out_path):
        with open(out_path, "w", encoding="utf-8") as f:
            for pred in preds:
                json.dump(pred, f, ensure_ascii=False)
                f.write("\n")
    else:
        print(f"Skipping {dataset} as {out_path} already exists.")    
    return

def acc_spd_test(bits_list=args.bits, residual_list=args.residual_length, group_size_list=args.group_size, batch_size_list=args.batch_size, datasets=args.datasets, model_name_or_path=args.model_name_or_path, cache_dir=args.cache_dir,out_path="./acc_spd_test"):
    bits_list = [int(x) for x in bits_list]
    residual_list = [int(x) for x in residual_list]
    group_size_list = [int(x) for x in group_size_list]
    batch_size_list = [int(x) for x in batch_size_list]
    config = LlamaConfig.from_pretrained(model_name_or_path)
    model_name = get_model_name.GetModelName(model_name_or_path)
    model2maxlen = json.load(open("config/model2maxlen.json", "r"))
    max_length = model2maxlen[model_name]
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    dataset2prompt = json.load(open("config/dataset2prompt.json", "r"))
    dataset2maxlen = json.load(open("config/dataset2maxlen.json", "r"))
    merged_test = []
    for bits in bits_list:
        for group_size in group_size_list:
            for residual_length in residual_list:
                SuccessMove = True
                if bits == 16 and residual_length == 32:
                    continue
                # load model and tokenizer
                config.k_bits = bits
                config.v_bits = bits
                config.group_size = group_size
                config.residual_length = residual_length
                if bits < 16:
                    from models.llama_kivi import LlamaForCausalLM_KIVI
                    config.use_flash = True
                    model = LlamaForCausalLM_KIVI.from_pretrained(
                        pretrained_model_name_or_path=model_name_or_path,
                        config=config,
                        cache_dir=cache_dir,
                        torch_dtype=torch.float16,
                        low_cpu_mem_usage=True,
                        device_map="auto",
                    )
                else:
                    from transformers import LlamaForCausalLM
                    model = LlamaForCausalLM.from_pretrained(
                        pretrained_model_name_or_path=model_name_or_path,
                        config=config,
                        cache_dir=cache_dir,
                        torch_dtype=torch.float16,
                        low_cpu_mem_usage=True,
                        device_map="auto",
                        attn_implementation="flash_attention_2"
                    )
                print(f"Begin to move model:{model_name}\ngroup_size:{group_size}, bits:{bits}, residual_length:{residual_length}")
                try:
                    model.to(device)
                except Exception as e:
                    print(f"The error is {e}")
                    SuccessMove = False
                if not SuccessMove:
                    continue
                print(device)
                tokenizer = AutoTokenizer.from_pretrained(
                        model_name_or_path, 
                        use_fast=False, 
                        trust_remote_code=True, 
                        tokenizer_type='llama')
                model.cuda().eval()
                tokenizer.pad_token = tokenizer.eos_token
                model.config.pad_token_id = model.config.eos_token_id
                tokenizer.padding_side = "left"
                for batch_size in batch_size_list:
                    is_outlier = False
                    test = {'model': model_name, 'bits': bits, 'group_size': group_size, 'residual_length': residual_length, 'batch_size': batch_size}

                    # load dataset
                    for dataset in datasets:
                        if dataset not in ["triviaqa", "qasper", "trec", "samsum", "lcc", "repobench-p", "qmsum", "multi_news"]:
                            continue
                        jsonl_path = f"{out_path}/{model_name}_{max_length}_{bits}bits_group{group_size}_residual{residual_length}_bs{batch_size}/{dataset}.jsonl"
                        test['dataset'] = dataset
                        print(f"Begin to load dataset {dataset}")
                        data = load_dataset('THUDM/LongBench', dataset, split='test')
                        print(f"Finish loading dataset {dataset}")
                        num_samples = len(data)
                        if num_samples < batch_size:
                            continue
                        prompt_format = dataset2prompt[dataset]
                        max_gen = dataset2maxlen[dataset]
                        test['max_gen'] = max_gen
                        prompts = []
                        prompt_batch = []
                        total_time = 0.0
                        token_number_list = []
                        output_token_number_list = []
                        print(f"bs: {batch_size}, bits:{bits}, dataset:{dataset}\nmodel:{model_name}")
                        preds = []
                        json_obj_list = []
                        for index, json_obj in tqdm(enumerate(data), total=num_samples):
                            json_obj_list.append(json_obj.copy())
                            prompt = prompt_format.format(**json_obj)
                            tokenized_prompt = tokenizer(prompt, truncation=False, return_tensors="pt").input_ids[0]
                            if len(tokenized_prompt) > max_length:
                                half = int(max_length/2)
                                prompt = tokenizer.decode(tokenized_prompt[:half], skip_special_tokens=True)+tokenizer.decode(tokenized_prompt[-half:], skip_special_tokens=True)
                                token_number_list.append(max_length)
                            else:
                                token_number_list.append(len(tokenized_prompt))
                            # Add this prompt to the batch.
                            prompt_batch.append(prompt)
                            # If the batch is not full and it's not the last prompt, continue adding prompts.
                            if len(prompt_batch) < batch_size and index < num_samples - 1: 
                                continue
                            # If the batch containing the last prompt is not full (i.e., less than the batch size), do not calculate the performance for this batch.
                            elif len(prompt_batch) < batch_size and index == num_samples -1:
                                token_number_list = token_number_list[:-len(prompt_batch)]
                                break
                            input = tokenizer(prompt_batch, padding=True, truncation=False, return_tensors="pt").to(device)
                            input_ids = input['input_ids']
                            context_length = input_ids.shape[-1]
                            torch.cuda.reset_peak_memory_stats()
                            with torch.no_grad():
                                torch.cuda.synchronize()
                                st = time.time()
                                if dataset == "samsum":
                                    try:
                                        output = model.generate(
                                            **input,
                                        max_new_tokens=max_gen,
                                        num_beams=1,
                                        do_sample=False,
                                        temperature=1.0,
                                        min_length=context_length+1,
                                        eos_token_id=[tokenizer.eos_token_id, tokenizer.encode("\n", add_special_tokens=False)[-1]],
                                        )
                                    except Exception as e:
                                        print(f"The error is {e}")
                                        is_outlier = True
                                else:
                                    try:
                                        output = model.generate(
                                            **input,
                                            max_new_tokens=max_gen,
                                            num_beams=1,
                                            do_sample=False,
                                            temperature=1.0,
                                        )
                                    except Exception as e:
                                        print(f"The error is {e}")
                                        is_outlier = True
                                if is_outlier == True:
                                    break
                                et = time.time()
                                infer_time = et - st;
                                total_time += infer_time
                                for i in range(batch_size):
                                    pred = tokenizer.decode(output[i][context_length:], skip_special_tokens=True)
                                    pred = post_process(pred, model_name)
                                    output_token_number_list.append(len(output[i][context_length:]))
                                    pred_index = index - batch_size + i + 1
                                    preds.append({"pred": pred, "answers": json_obj_list[pred_index]["answers"], "all_classes": json_obj_list[pred_index]["all_classes"], "length": json_obj_list[pred_index]["length"]})
                                torch.cuda.empty_cache()
                                torch.cuda.synchronize()
                            # Successfully executed, then add the prompt to prompts.
                            prompts.append(prompt)
                            # Clear the prompt_batch.
                            prompt_batch = []
                        if is_outlier == True:
                            break
                        assert len(token_number_list) == num_samples//batch_size*batch_size
                        pred2jsonl(jsonl_path, dataset, preds)
                        test['batch_num'] = num_samples//batch_size
                        test['used time'] = total_time * 1000
                        # print(sum(token_number_list))
                        test['throughput'] = test['max_gen']*test['batch_num']*test['batch_size']*1000/test['used time']
                        test['throughput2'] = sum(output_token_number_list)*1000/test['used time']
                        print(f"max_gen: {test['max_gen']}, batch_num: {test['batch_num']}, batch_size: {test['batch_size']}, Throughput1: {test['throughput']}\n sum(token_number_list): {sum(output_token_number_list)}, Throughput2: {test['throughput2']}")
                        merged_test.append(test.copy())
                    if is_outlier == True: # The batch size is too large. Let's proceed with testing the next batch size.
                        continue
                del model
                torch.cuda.empty_cache()
    return model_name, merged_test

cwd = os.getcwd()
out_path = os.path.join(cwd, "acc_spd_test")
os.makedirs(out_path, exist_ok=True)
model_name, merged_test = acc_spd_test(out_path=out_path)
wb = Workbook()
ws = wb.active
if len(merged_test)==0:
    print("No data")
    import sys
    sys.exit()
headers = list(merged_test[0].keys())
ws.append(headers)

for test in merged_test:
    data = [test[key] for key in headers]
    ws.append(data)

datasets_string = '_'.join(args.datasets)

if len(args.bits) == 1:
    wb.save(f'{out_path}/acc_speed_{model_name}_KIVI{args.bits[0]}_{datasets_string}.xlsx')
else:
    if len(args.batch_size) == 1:
        wb.save(f'{out_path}/acc_speed_{model_name}_bs{args.batch_size[0]}_{datasets_string}.xlsx')
    else:
        wb.save(f'{out_path}/acc_speed_{model_name}_{datasets_string}.xlsx')`

**Comment by Felixvillas** [comment]
> Thank you for the information.
> 
> > Is there a shared state or cache within a batch that could be causing this issue?
> 
> No, there is no sharing across batch size. In KIVI, for key cache, scale and offset factor are shared between different tokens within the same sequence; for value cache, scale and offset factor are shared between different channels within the same token embedding.
> 
> It is a bug. My guess is because of padding, which pad different request into the same length. The current [pred_long_bench.py](https://github.com/jy-yuan/KIVI/blob/main/pred_long_bench.py) does not support inference with batch size > 1. Can you share your test file & script with me? So I can have a try.

I rethought it, since it's the padding part causing the inaccurate KIVI precision, why doesn't this problem occur when `bits=16`?

**Comment by zirui-ray-liu** [comment]
> I rethought it, since it's the padding part causing the inaccurate KIVI precision, why doesn't this problem occur when bits=16?

Because KIVI and fp16 baseline has completely separate attention implementation. In KIVI we do not carefully examine the code regarding the [attention mask](https://github.com/jy-yuan/KIVI/blob/main/models/llama_kivi.py#L390).Recently I have been very busy. I will review the script you provided once I got some free time

**Comment by Felixvillas** [comment]
> > I rethought it, since it's the padding part causing the inaccurate KIVI precision, why doesn't this problem occur when bits=16?
> 
> Because KIVI and fp16 baseline has completely separate attention implementation. In KIVI we do not carefully examine the code regarding the [attention mask](https://github.com/jy-yuan/KIVI/blob/main/models/llama_kivi.py#L390).Recently I have been very busy. I will review the script you provided once I got some free time

Thanks, I'm  looking forward to it.
